lib/src/rules/use_build_context_synchronously.dart

```diff
@@ -0,0 +1,199 @@
+
+import '../analyzer.dart';
+
+// todo (pq): flesh out storing context gotchas -- codify here or in another lint?
+
+const _desc = r'Do not use BuildContexts across async calls.';
+
+const _details = r'''
+**DO NOT** use BuildContexts across async calls.
+
+TODO: add rationale.
+
+TODO: describe mechanics.
+
+
+**BAD:**
+```
+class MyState extends State<MyWidget> {
+  void m() async {
+    // Uses context from State.
+    Navigator.of(context).pushNamed('routeName');
+
+    await Future<void>.delayed(Duration());
+    // ^--- async gap.
+
+    // Without a mounted check, this is unsafe.
+    Navigator.of(context).pushNamed('routeName'); // LINT
+  }
+}
+```
+
+**GOOD:**
+```
+class MyState extends State<MyWidget> {
+  void m() async {
+    // Uses context from State.
+    Navigator.of(context).pushNamed('routeName');
+
+    await Future<void>.delayed(Duration());
+
+    if (!mounted) return;
+
+    // OK. We checked mounted checked first.
+    Navigator.of(context).pushNamed('routeName'); // OK
+  }
+}
+```
+''';
+
+class UseBuildContextSynchronously extends LintRule implements NodeLintRule {
+  UseBuildContextSynchronously()
+      : super(
+            name: 'use_build_context_synchronously',
+            description: _desc,
+            details: _details,
+            group: Group.errors,
+            maturity: Maturity.experimental);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    final visitor = _Visitor(this);
+    registry.addMethodInvocation(this, visitor);
+    registry.addInstanceCreationExpression(this, visitor);
+    registry.addFunctionExpressionInvocation(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  static const _nameBuildContext = 'BuildContext';
+  final Uri _uriFramework;
+
+  // todo (pq): replace in favor of flutter_utils.isBuildContext
+  final LintRule rule;
+
+  _Visitor(this.rule)
+      : _uriFramework = Uri.parse('package:flutter/src/widgets/framework.dart');
+
+  bool accessesContext(ArgumentList argumentList) {
+    for (var argument in argumentList.arguments) {
+      var argType = argument.staticType;
+      if (isBuildContext(argType)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void check(AstNode node) {
+    // Walk back and look for an async gap that is not guarded by a mounted
+    // property check.
+    var child = node;
+    while (child != null && child is! FunctionBody) {
+      var parent = child.parent;
+      if (parent is Block) {
+        var statements = parent.statements;
+        var index = statements.indexOf(child as Statement);
+        for (var i = index - 1; i >= 0; i--) {
+          var s = statements.elementAt(i);
+          if (isMountedCheck(s)) {
+            return;
+          } else if (isAsync(s)) {
+            rule.reportLint(node);
+            return;
+          }
+        }
+      }
+
+      child = child.parent;
+    }
+  }
+
+  bool isAsync(Statement statement) {
```

Comment:
The idea is right, but I think you actually need to use a full visitor here to visit every part of the statement. The `await` expression doesn't need to be the immediate child of an expression statement. Consider, for example, `print(await f());`. This method will miss that case.

Reply:
Absolutely.  Ironically, that's much simpler too.  Done.